### ISSUE TITLE
URI upper/lower case

### ISSUE THREAD

**Issue by wolispace** [putyourlightson/craft-blitz#623]
### Support Request

All of our slugs are lowercase, yet I find I can enter uppercase letters and they are stored in the cache table as typed:
```
MariaDB [db]> select id, uri from blitz_caches where uri like '%about-us%' order by uri;
+----+----------------------+
| id | uri                  |
+----+----------------------+
|  8 | ABOUT-us             |
| 13 | about-us/Legislation |
| 11 | ABOUT-us/OUR-goals   |
+----+----------------------+
3 rows in set (0.003 sec)
```

The pages load as expected, no issues there.

The folders in web/cache/blitz/.. are all lowercase as expected...

... yet somehow an uppercase version slipped through so we had two folders:

- about-us/Legislation/index.html
- about-us/legislation/index.html

And using the different cased URI shows the different cached content.

I have yet to replicate this in my offline copy, but thought I would raise it in case it's something you know about.

Could a rebuild of catchable content read the URI  from the DB and use that to re-generate the folders?

Thanks
Wallace

**Comment by bencroker** [comment]
URIs are generally case sensitive, depending on the web server, so `about-us` and `ABOUT-us` can indeed be unique.

When regenerating content, Blitz references pages previously cached, so if you want to rebuild the cache without those unexpected URLs then _flushing_ the cache before forcing a refresh will clear out previously cached URIs in the database.

**Comment by wolispace** [comment]
Craft appears to be auto-converting the slugs into lowercase.

Could Blitz do a similar thing?

When I visit about-us but use some capitals I get this on disk:
```
$ find . -type d -print | sed -e 's;[^/]*/;|____;g;s;____|; |;g'
.
|_____includes
| |____action=blitz%2Finclude%2Fcached&index=1226925700
| |____action=blitz%2Finclude%2Fcached&index=199726758
| |____action=blitz%2Finclude%2Fcached&index=2063942130
| |____action=blitz%2Finclude%2Fcached&index=891906451
| |____action=blitz%2Finclude%2Fcached&index=3793990563
|____About-Us
```
When I then visit the same page but with all lowercase I get a new page being cached:
```
$ find . -type d -print | sed -e 's;[^/]*/;|____;g;s;____|; |;g'
.
|_____includes
| |____action=blitz%2Finclude%2Fcached&index=1226925700
| |____action=blitz%2Finclude%2Fcached&index=199726758
| |____action=blitz%2Finclude%2Fcached&index=2063942130
| |____action=blitz%2Finclude%2Fcached&index=891906451
| |____action=blitz%2Finclude%2Fcached&index=3793990563
|____About-Us
|____about-us
```

The `blitz_caches` table only stores the last version of the uri, so first is was 'About-Us' and then it changes to 'about-us'.

Since both come from the one about-us page, editing that will only refresh one of those index.html files (the last one in `blitz_caches` I assume). 

Someone visiting with About-Us will get that cached content that may not be the most up-to-date.

**Comment by bencroker** [comment]
I think the underlying question here is, should `About-Us` be a valid URI? As previously stated, URIs are case-sensitive, so `about-us` and `About-Us` are _not_ the same page. My guess is that if you track down why uppercase letters are not returning 404 responses, then you’ll be able to fix the root issue.

**Comment by wolispace** [comment]
I have found we are using the default Craft config for DB collation:
https://craftcms.com/docs/4.x/config/db.html#collation

So `AbOuT-uS` and `about-us` are the same thing as far as Craft queries are concerned (in a default configuration).

Is it possible Blitz can query the DB to find out if a "*_ci" case insensitive collation has been used and lowercase the directory names when creating the cache?

Or maybe give us a config setting to enable this as needed?

**Comment by bencroker** [comment]
I’ll need to think about this some more. It seems like it could be handled at the web server level, if it is an issue. For example, you could force all URIs to be rewritten to lowercase before they even reach Craft.

**Comment by wolispace** [comment]
Yes, I thought about that. Only lowercase Craft requests. Seems like a complex setup process for everyone who is using default Craft and Blitz.   And if anyone hosts multiple Craft instances and allows customers to install plugins like Blitz, then there maybe friction trying to ask the host to modify their web server config just for them.

I also through about clearing cached content on disk on a more regular basis, but you can still get in a muddle with multiple cached content on disk point to the one Craft entry so only one gets the most up-to-date content until all cached content is cleared. Which defeats the purpose of a cache.

Personally I would prefer a way of forcing lowercase directory names in blitz.

I'll let you ponder some more.

**Comment by bencroker** [comment]
I thought about this in detail and am reluctant that Blitz should do any case conversion. While Craft converts slugs to lower case, it is still possible to have template routes and custom routes that contain upper case characters. I don’t see any reason why Blitz should not cache these. And basing it off of the database collation is not an option, since individual columns and individual tables can be assigned different collations.

My recommendation for dealing with upper case characters is going to depend on what you want to achieve.

1. If you want your site to only allow lower case URIs, then add a server rewrite to convert upper case characters in URIs to lower case.
2. If you want Blitz to only cache lower case URIs, then add an **excluded URI pattern** to prevent the caching of URIs containing upper case characters using the following regular expression, for example:
    ```
    [A-Z]+
    ```
    Note that the regular expression above only includes upper case characters in English, so you may need to manually add other characters.
    ```
    [A-ZÄÖÜ]+
    ```

**Comment by wolispace** [comment]
Ah.. excluding the malformed urls from Blitz cache makes sense.

so blitz.php would contain something like:
```
'excludedUriPatterns' => [
            [
                'siteId' => 1,
                'uriPattern' => '[A-Z]+',
            ],
        ],
```
It looks like a pattern of /\p{Lu}+/gu will match any unicode uppercase character.. in php:
`$pattern = '/\p{Lu}+/u'; // 'u' flag for Unicode matching`

Would this work in the code you use for pattern matching or do we need to use something like this (thanks Copilot):
[A-ZÄÖÁÉÍÓÚÀÈÌÒÙÂÊÎÔÛÃÑÕÇËÜÏÅÆŒ]+

**Comment by bencroker** [comment]
Yes, that config setting looks right. Using `\p{Lu}+` should match any upper case unicode character, but I have yet to use it.

```php
'excludedUriPatterns' => [
    [
        'siteId' => 1,
        'uriPattern' => '\p{Lu}+',
    ],
],
```

**Comment by wolispace** [comment]
I can confirm that pattern works. 

_(I edited my blitz.php, then visited the /settings/plugins/blitz and clicked Save.. however I could have just entered it on that settings page I see now)._

When I go to a page, replacing some characters in the slug with uppercase, and/or uppercase with inflections, it returns the page I want as if I had typed the slug in all lowercase, and we see the log has skipped them:

```
[2024-03-04 13:17:07] Page not cached because it matches an excluded URI pattern. [https://mysite.test/about-us/OUR-goals]
[2024-03-04 13:18:09] Page not cached because it matches an excluded URI pattern. [https://mysite.test/about-us/OÜR-goals]
```

Because the page returned is not cached (slow - but thats the price for incorrect slugs), it has all the latest edits, just like the cached version that gets updated when the page is edited and cache refreshed.

All good. Thanks.
